In [26]:
import pickle

def load_obj(name ):
    """ load object """
    with open('obj/' + name + '.pkl', 'rb') as f:
        return pickle.load(f)

master_bizdet_df = load_obj('master_bizdet_df2')
my_rev_features = load_obj('my_rev_features')
my_bm_features = load_obj('my_bm_features')
friend_rev_features = load_obj('friend_rev_features')
friend_bm_features = load_obj('friend_bm_features')
both_sim = load_obj('both_sim')
rec_details = load_obj('rec_details')

# Yelp Recommendation Engine: 

## Background:
Yelp is a local-search serviced powered by crowd-sourced reviews of local businesses. I am a frequent visitor of Yelp and rely heavily on the reviews to make decisions on which restaurants to visit, what to order at the restaurant, and outside of food, I also rely on Yelp reviews for services I seek. In addition to leaving reviews with a rating between 1-5, users with an account with Yelp are able to add friends and to create Bookmarks for businesses they want to visit in the future and can refer to later. Yelp also has their own API has various endpoints that provide different information related to businesses, events, and categories.

## Purpose: 
To build a recommendation system for Yelp using bookmarks and review ratings and the users that are most similar to me.

I spend a lot of time going through Yelp trying to find a place to eat, especially if I am not craving anything in particular. With a recommendation engine, I can narrow down the list of places that I may enjoy.

## Acquiring Data Set: 

#### Scraped
The following data will be scraped from www.Yelp.com:
- list of friends of my account - `user_id`
- list of my bookmarks - `business_id`, `url`
- list of my reviews - `business_id`, `url`, `user_rating`
- list of my friend's bookmarks - `business_id`, `url` 
- list of my friend's reviews - `business_id`, `url`, `user_rating`

#### API
For each business in the above, data from Business Details Endpoint from __[Yelp Fusion API](https://www.yelp.com/developers/documentation/v3/business)__ :

Note: Yelp's API has a limit of 5000 calls per day

Relevant fields:

| Name | Type   |  Description   |
|------|------|------|
|   categories | object[]| A list of category title and alias pairs associated with this business.|
|   categories[x].alias | string| Alias of a category, when searching for business in certain categories, use alias rather than the title.|
|   alias | string| Unique Yelp alias of this business. Can contain unicode characters. Example: 'yelp-san-francisco'. Also see: What's the difference between the Yelp business ID and business alias?|
|   is_closed | bool| Whether business has been (permanently) closed|
|   price | string| Price level of the business|
|   rating | decimal| Rating for this business|
|   review_count | int| Number of reviews for this business.|
|   id | str| Unique Yelp ID of this business.|

## Cleaning Data and Feature Engineering

1. `master_bizdet_df` For each business, create a dataframe containing business details. 
2. For my own bookmarks + reviews and my friend's bookmarks + reviews, create a dataframe of features that will be used to calculate similarities
    - `price_val` - average price level for businesses bookmarked/reviewed (1 being cheapest, 4 being priciest)
    - `rating` - average rating of businesses bookmarked/reviewed (1 being lowest, 5 being highest)
    - `review_count` - number of reviews the business has
    - `categories` - the column category's percentage of total bookmarks/reviews. Note that a business can have up to 4 different categories

In [9]:
master_bizdet_df.head() # relevant business details

alias        cat_0            cat_1  \
0          fork-in-aussie-pies-santa-monica-2       coffee       australian   
0                    providence-los-angeles-2      seafood              NaN   
0  society-billiards-and-cafe-pacific-beach-3    poolhalls             bars   
0            island-vintage-coffee-honolulu-4       coffee              NaN   
0      beat-the-lock-escape-rooms-santa-clara  escapegames  kids_activities   

  cat_2 cat_3 cat_4                      id  is_closed price  rating  \
0   NaN   NaN   NaN  5lJMJRTui0tQ3G8gdEHo7A       True    $$     4.5   
0   NaN   NaN   NaN  TzIJzamxdVGc3zReKbLGaA      False  $$$$     4.5   
0   NaN   NaN   NaN  _nEYB0urvuAxSBrFpYxwcg       True    $$     3.0   
0   NaN   NaN   NaN  UdvXV2ux3uOj3UqP04cjqA      False    $$     4.5   
0   NaN   NaN   NaN  0qj734IaYhGgAg2jOpiI2g      False   NaN     4.5   

   review_count  price_val  
0           634        2.0  
0          2478        4.0  
0           154        2.0  
0          1317        2.0  
0            82        NaN

In [7]:
my_rev_features.head() # features used to calculate similarity

price_val    rating  guamanian  eyelashservice  \
user_id                                                                  
ShHBKjuJbQAVBLs7DgA95A   2.072727  4.064516   0.014925        0.014925   

                        japanese  newamerican   seafood     pizza      thai  \
user_id                                                                       
ShHBKjuJbQAVBLs7DgA95A  0.134328     0.044776  0.119403  0.014925  0.074627   

                        mobilephones ...   tradamerican   museums  \
user_id                              ...                            
ShHBKjuJbQAVBLs7DgA95A      0.014925 ...       0.014925  0.014925   

                        othersalons     tapas  desserts   buffets  \
user_id                                                             
ShHBKjuJbQAVBLs7DgA95A     0.014925  0.014925  0.014925  0.014925   

                        laserlasikeyes  musicvenues  sportsbars  num  
user_id                                                               
ShHBKjuJbQAVBLs7DgA95A        0.014925     0.014925    0.014925  1.0  

[1 rows x 73 columns]

In [29]:
print('my bookmark features shape: ', my_bm_features.shape)
print('my review features shape: ', my_rev_features.shape)
print('friends bookmark features shape: ', friend_bm_features.shape)
print('friends review features shape: ', friend_rev_features.shape)

my bookmark features shape:  (1, 124)
my review features shape:  (1, 73)
friends bookmark features shape:  (144, 693)
friends review features shape:  (59, 554)


## Exploring Data

Looks like my friends and I have high standards for reviews and for bookmarks but like to eat at cheaper places. At least 75% of all businesses had a rating of 4 stars or more and has 2 dollar signs or less.

In [10]:
master_bizdet_df.describe()

rating  review_count     price_val
count  16707.000000  16707.000000  15074.000000
mean       4.060962    561.238702      1.857967
std        0.487062    776.005924      0.676829
min        1.000000      1.000000      1.000000
25%        4.000000    121.000000      1.000000
50%        4.000000    314.000000      2.000000
75%        4.500000    701.000000      2.000000
max        5.000000  15746.000000      4.000000

Now I want to see what are the top 10 most common categories among all users that were bookmarked or reviewed and compare it to my own bookmarks and my own reviews.

<img src="./Yelp Pics/All Categories Bar Chart.png">
<img src="./Yelp Pics/My Reviews_Category Bar Chart.png">
<img src="./Yelp Pics/My Bookmarks_Category Bar Chart.png">

I definitely see similarities among the categories already. Looks like we all enjoy coffee, breakfast/brunch, sushi, japanese, seafood, and bakeries.

## Calculate Similarity

In order to determine the similarity between myself and the other users, I am calculating the similarity of bookmarks and reviews separately. I am assuming that bookmarks represent businesses the user wants to visit but has not yet visited and that reviews are businesses that the user has already been to. With that, I am treating that as another factor of similarity (whether or not we have been to the business). After I calculate the similarities of both reviews and bookmarks, I will average them together to determine a final similaritiy and take the top 3 users to get recommendations

In [15]:
both_sim[:10] # top 10 similar users

bm_similarity  rev_similarity     total       avg
ShHBKjuJbQAVBLs7DgA95A       1.000000        1.000000  2.000000  1.000000
WJfNVm4mXXDt1Vakg13CpA       0.999580        0.998110  1.997691  0.998845
QjLiYeQLeMIqYu-ncmkUXg       0.999154        0.998428  1.997582  0.998791
oLh6PxTxdMAufjerdyZz5w       0.998793        0.998431  1.997223  0.998612
5RifcJP_Lf-MzojTHybBNw       0.999305        0.997782  1.997086  0.998543
7IV3JXTXufasWNSkDEdrcQ       0.998290        0.998115  1.996405  0.998202
NmdKCZlUVJGnqDxhXl2ofw       0.999074        0.997216  1.996291  0.998145
a-Y3kWoP_HD_lBjD7qdfZg       0.998097        0.997892  1.995990  0.997995
MzC1_5kXxGw336fMYSrsdg       0.999540        0.996266  1.995805  0.997903
Zw-0Lo01W0QdkHT60DYg4g       0.998985        0.996579  1.995564  0.997782

## Get Recommendations

To get recommendations:
1. Start with all the IDs reviews and bookmarks of the three users
2. Remove the businesses that I have already bookmarked or reviewed
3. Filter for businesses that all three users have bookmarked or reviewed
4. Filter out businesses that are closed (`is_closed` == True)
5. Create a dataframe with the resulting business IDs and merge with the master business detail dataframe

In [21]:
rec_details

alias             cat_0  \
0                  b-patisserie-san-francisco-2          bakeries   
0         billys-boston-chowder-house-los-gatos           seafood   
0                        bon-nene-san-francisco          japanese   
0                       cockscomb-san-francisco              bars   
0                         coqueta-san-francisco           spanish   
0            craftsman-and-wolves-san-francisco          bakeries   
0                            dishdash-sunnyvale        mideastern   
0  forma-restaurant-and-cheese-bar-santa-monica           italian   
0                         frances-san-francisco       newamerican   
0                               gather-berkeley       newamerican   
0                             gregoire-berkeley            french   
0                           hash-house-campbell  breakfast_brunch   
0      humphry-slocombe-ice-cream-san-francisco          icecream   
0                         in-situ-san-francisco       newamerican   
0                            la-note-berkeley-5            french   
0                  liguria-bakery-san-francisco          bakeries   
0                         mama-chef-santa-clara         taiwanese   
0             matcha-cafe-maiko-san-francisco-2            coffee   
0          mission-chinese-food-san-francisco-4           chinese   
0             mr-holmes-bakehouse-san-francisco          bakeries   
0                            nopa-san-francisco       newamerican   
0                    pesce-e-riso-san-francisco          japanese   
0                    pine-and-crane-los-angeles         taiwanese   
0                pono-hawaiian-grill-santa-cruz          hawaiian   
0                     seven-hills-san-francisco           italian   
0                         sushi-gen-los-angeles             sushi   
0                sushi-sams-edomata-san-mateo-2             sushi   
0                     sushi-yoshizumi-san-mateo             sushi   
0               swan-oyster-depot-san-francisco           seafood   
0                              takashi-new-york          japanese   
0         tartine-bakery-and-cafe-san-francisco          bakeries   
0                          the-cottage-la-jolla       newamerican   
0             the-penny-ice-creamery-santa-cruz          icecream   
0                  the-picnic-basket-santa-cruz  breakfast_brunch   
0                  the-progress-san-francisco-2       newamerican   
0           tomales-bay-oyster-company-marshall    seafoodmarkets   
0                         trestle-san-francisco       newamerican   
0                   tsubasa-sushi-san-francisco             sushi   
0                            vesta-redwood-city             pizza   
0                           viks-chaat-berkeley            indpak   
0                           wood-tavern-oakland       newamerican   

              cat_1             cat_2 cat_3 cat_4                      id  \
0          cakeshop          macarons   NaN   NaN  2XQm-uFcTS7oc8MFP-8olA   
0        fishnchips  breakfast_brunch   NaN   NaN  z0wcq5E8GgIUgc-O9JDEGQ   
0   tapasmallplates       asianfusion   NaN   NaN  RxIFKVvc8iVBPIVpe55Byg   
0       newamerican               NaN   NaN   NaN  oaBpGm6woGN69J9iRqu0Fw   
0   tapasmallplates               NaN   NaN   NaN  xv8dq_jBNfQkHeX2wJGlAg   
0             cafes          cakeshop   NaN   NaN  AfqpSxetSUMc63ZPCfbneg   
0     mediterranean               NaN   NaN   NaN  tRiDMqAxeBERRvuyYPZ-aw   
0       newamerican               NaN   NaN   NaN  S7VJq9qqXCCoiHEbwb3tCw   
0          desserts         wine_bars   NaN   NaN  dMlsQsPU8Y_oiPYC5hHi2A   
0  breakfast_brunch              bars   NaN   NaN  BiZAzf2SAFCE0gJrX-PGFw   
0        sandwiches               NaN   NaN   NaN  sKnjZ4bb-0AWxWqJv7cHOw   
0       newamerican               NaN   NaN   NaN  VxgMRQLmA0Lgjzmt2PG6Eg   
0               NaN               NaN   NaN   NaN  47OC_X6KkiDDQ4jwoCUjFg   
0           seafood               NaN   NaN   NaN  D8-lVk1RQPPAap

In [22]:
rec_details.describe()

rating  review_count  price_val
count  41.000000     41.000000  41.000000
mean    4.158537   1763.707317   2.219512
std     0.324789   1439.636382   0.790955
min     3.500000     98.000000   1.000000
25%     4.000000    738.000000   2.000000
50%     4.000000   1447.000000   2.000000
75%     4.500000   2174.000000   3.000000
max     4.500000   7464.000000   4.000000

<img src="./Yelp Pics/Recommendations Bar Chart.png">

## Results

The results are a list of 41 businesses that have a rating of 3.5 or higher and appear very popular with hundreds of reviews. While I didn't add a column to indicate location, the business name includes the location and looking through it, most of the businesses are located in the Bay Area.

## Limitations and Further Improvement

The recommendations does not remove businesses that I have been to but not reviewed or bookmarked as there is no check-in data available (but even if there was, the user needs to be consistently using the check-in feature)

I had used the business name instead of business ID to call the business details endpoint from Yelp Fusion. Because some business names have special characters, those businesses were unable to be joined with their business details. As a result, there are 1021 businesses that have null business details. To further improve, I would scrape Yelp.com for the business IDs and recall the business details endpoint using the business ID to have a more complete dataset. For time's sake, I determined that 1021/16707 or 6% of the data is small enough to be able to continue with the analysis.